File: app/src/main/java/com/expensivebelly/dagger2retaingraph/activity/RetainGraphActivity.java
package com.expensivebelly.dagger2retaingraph.activity;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.widget.TextView;

import com.expensivebelly.dagger2retaingraph.R;
import com.expensivebelly.dagger2retaingraph.activity.di.RetainGraphComponent;

import javax.inject.Inject;

/**
 * Dependency graph is retained through configuration changes in the Activity
 */

public class RetainGraphActivity extends AppCompatActivity implements RetainGraphView {

    @Inject
    RetainGraphPresenter presenter;

    private RetainGraphComponent component;
    private TextView messageView;

    public static Intent startIntent(Activity activity) {
        return new Intent(activity, RetainGraphActivity.class);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_retain_graph);
        messageView = (TextView) findViewById(R.id.text_message);

        component().inject(this);

        presenter.attachView(this);
    }

    private RetainGraphComponent component() {
        if (component == null) {
            component = (RetainGraphComponent) getLastCustomNonConfigurationInstance();
            if (component == null) {
                component = RetainGraphComponent.Initializer.init();
            }
        }
        return component;
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        presenter.detachView();
    }

    @Override
    public Object onRetainCustomNonConfigurationInstance() {
        return component();
    }

    @Override
    public void display(String message) {
        messageView.setText(message);
    }
}

File: app/src/main/java/com/expensivebelly/dagger2retaingraph/activity/RetainGraphActivity.kt
package com.expensivebelly.dagger2retaingraph.activity

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import com.expensivebelly.dagger2retaingraph.R
import com.expensivebelly.dagger2retaingraph.activity.di.DaggerRetainGraphComponent
import com.expensivebelly.dagger2retaingraph.activity.di.RetainGraphComponent
import com.expensivebelly.dagger2retaingraph.activity.di.RetainModule
import javax.inject.Inject

/**
 * Dependency graph is retained through configuration changes in the Activity
 */
class RetainGraphActivity : AppCompatActivity(), RetainGraphView {

    @Inject
    lateinit var presenter: RetainGraphPresenter

    private val component by lazy {
        (lastCustomNonConfigurationInstance as RetainGraphComponent?)
                ?: DaggerRetainGraphComponent.builder().build()
    }

    private val messageView by lazy {
        findViewById<TextView>(R.id.text_message)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_retain_graph)
        component.inject(this)
    }

    override fun onStart() {
        super.onStart()
        presenter.attach(this)
    }

    override fun onStop() {
        if (!isChangingConfigurations) presenter.detach()
        super.onStop()
    }

    override fun onRetainCustomNonConfigurationInstance() = component

    override fun display(message: String) {
        messageView.text = message
    }

    companion object {
        fun startIntent(activity: Activity) = Intent(activity, RetainGraphActivity::class.java)
    }
}